feat: data warehouse sources access control

zlwaterfield opened the pull request:
## Changes


Adding access control for data warehouse sources. A user can control access to a single source or all sources across the project. Includes client and server checks. 

See docs
- https://posthog.com/docs/settings/access-control
- https://github.com/PostHog/posthog/blob/master/.claude/agents/access-control.md

It's not perfect, few missing parts
- if a user has view only access to all sources and they land on http://localhost:8010/project/1/data-warehouse/new-source?kind=azure they will be able to use the form but the server request will fail
- it does not limit querying the source yet, ideally if a user as no acccess to a source they shouldn't be able to query it - if this is a problem we could not allow the no access option for now or show a warning. 
- this is only for managed sources and the external data source model

<img width="512" height="646" alt="Screenshot 2025-10-07 at 3 59 57 PM" src="https://github.com/user-attachments/assets/a73180a5-c514-4126-9058-9a146462211f" />

<img width="612" height="244" alt="Screenshot 2025-10-07 at 4 00 24 PM" src="https://github.com/user-attachments/assets/51c28870-85d4-40c1-ab19-905ee2ac8fe4" />

<img width="553" height="433" alt="Screenshot 2025-10-07 at 4 09 38 PM" src="https://github.com/user-attachments/assets/0f312d3f-7ad1-4c99-a136-d20e5d809435" />

## How did you test this code?

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

<!-- Docs reminder: If this change requires updated docs, please do that! Engineers are the primary people responsible for their documentation. 🙌 -->

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Changelog: (features only) Is this feature complete?

<!-- Optional, but helpful for our content team! -->
<!-- Yes if this is okay to go in the changelog. No if it's still hidden behind a feature flag, or part of a feature that's not complete yet, etc.  -->

zlwaterfield commented on posthog/warehouse/api/external_data_source.py:
```diff
@@ -285,12 +289,12 @@ class Meta:
     """
 
-    scope_object = "INTERNAL"
+    scope_object = "external_data_source"
```
We won't add to scopes for personal API keys

zlwaterfield committed:
initial commit for data warehouse sources

a-lider reviewed (changes requested):
I think the main use case for data warehouse access control is to restrict someone from creating/editing data sources and querying existing data warehouse sources.

Might make sense to reframe the PR as 'Sources access control', then handle SQL editor access controls separately (blocking any custom queries, including joins with data warehouse tables).
This way, Sources access control could include Event sources + Managed and Self-managed data warehouse sources, not just data warehouse sources. 

Basically all of these sections: 

<img width="1613" height="1004" alt="Screenshot 2025-10-09 at 13 56 06" src="https://github.com/user-attachments/assets/18890392-7c04-4712-b679-8d71e41ac111" />

Open to discuss, just an idea.

a-lider reviewed (commented):
Posting here for the record, chatted about this on standup. Makes sense to keep it as "Data warehouse sources" access control and cover both Managed and Self-managed sources. 

We still need to add a no-access state for the [Data management -> Sources](http://localhost:8010/project/6/pipeline/sources) page. I can finish this if you're short on time. 
<img width="1070" height="494" alt="Screenshot 2025-10-14 at 15 46 27" src="https://github.com/user-attachments/assets/eef9c902-ea74-4816-8edc-a17b33b7c905" />

zlwaterfield committed:
clean up some ordering

zlwaterfield committed:
Update HogFunctionTemplateList.tsx

github-actions[bot] committed:
Update query snapshots

zlwaterfield committed:
Add tooltips for resource-specific access control info

Introduces the getAccessControlTooltip utility to provide contextual tooltips for resources with special access control behavior. Updates AccessControlObject and ResourcesAccessControls components to display an info icon with a tooltip when relevant, improving user understanding of resource-specific access control limitations.

zlwaterfield committed:
Add access control to data warehouse UI actions

Introduced access control checks and UI gating for data warehouse sources and schema management. Actions such as editing, deleting, syncing, and updating sources and tables are now wrapped in AccessControlAction components, ensuring only users with sufficient permissions can perform them. Also updated types and tests to support user access levels.

zlwaterfield committed:
LINT / tests

Gilbert09 reviewed (approved):
One nit

andrewjmcgehee commented:
@Gilbert09 are you taking over this PR since Zach left PostHog? if so, noting here that the customer in [this ticket](https://posthoghelp.zendesk.com/agent/tickets/40853) should be notified when this merges because they specifically were after access controls for MySQL sources. noting here so we can close the ticket. i'm also going to post a link to this PR in the ticket for the customer so they can keep tabs on the progress themselves.

yasen-posthog committed:
fix: Add ProductTour back to AccessControlResourceType and update query count

github-actions[bot] committed:
test(backend): update query snapshots

yasen-posthog committed:
feat(data-warehouse): add access control tests and improve self-managed source detection

- Add comprehensive access control test suite (24 tests) for ExternalDataSource
- Add `isSelfManagedSource` selector to sourceWizardLogic for explicit detection
- Add `isManagedSourceTemplate` utility for consistent managed/self-managed checks
- Fix 403 handling in externalDataSourcesLogic to return empty list
- Update HogFunctionTemplateList to use shared utility function

yasen-posthog committed:
fix(data-warehouse): add external_data_source to APIScopeObject type

yasen-posthog committed:
fix(data-warehouse): exclude self-managed sources from access control

yasen-posthog committed:
refactor(data-warehouse): add unified source ID utility functions

yasen-posthog committed:
fix(data-warehouse): remove scene-level access control for self-managed sources

yasen-posthog committed:
refactor(data-warehouse): simplify AccessControlAction usage in Schemas

- Add SourceEditorAction helper component to reduce repetitive props
- Export AccessControlActionChildrenProps for reuse
- Remove redundant nested AccessControlAction wrappers in actions column
- Reduces code by ~33 lines while maintaining same functionality

yasen-posthog committed:
fix(data-warehouse): only suppress 403 errors when loading available sources

Return null only for 403 (user lacks access), rethrow all other errors.

yasen-posthog committed:
chore(data-warehouse): regenerate OpenAPI types for user_access_level

a-lider reviewed (approved):
Looks good. I think its fine to release without self-managed sources and add them later
